pkg/sql/pg_catalog_test.go

```diff
@@ -0,0 +1,379 @@
+)
+
+// This constants enumerates indexes in pg_catalog_tables
+const (
+	testTableIndex = iota
+	testColumnIndex
+	testTypeIndex
+	testTypeOidIndex
+)
+
+// This query uses udt_name::regtype instead of data_type column because
+// data_type only says "ARRAY" but does not say which kind of array it is.
+const getPgCatalogSQL = `
+	SELECT table_name, column_name, udt_name::regtype data_type, udt_name::regtype::Oid data_type_oid
+	FROM information_schema.columns 
+	WHERE table_schema = 'pg_catalog'
+	ORDER BY 1, 2;
+`
+
+// Test data files
+const (
+	pgCatalogDump = "pg_catalog_tables"          // PostgreSQL pg_catalog schema
+	expectedDiffs = "pg_catalog_test-diffs.json" // Contains expected difference between postgres and cockroach
+	testdata      = "testdata"                   // testdata directory
+)
+
+// When running test with -rewrite-diffs test will pass and re-create pg_catalog_test-diffs.json
+var rewriteFlag = flag.Bool("rewrite-diffs", false, "This will re-create the expected diffs file")
+
+// summary will keep accountability for any unexpected difference and report it in the log
+type summary struct {
+	missingTables        int
+	missingColumns       int
+	mismatchDatatypesOid int
+}
+
+// TestColumn is a structure which contains a small description about the datatype of a column, but this can also be
+// used as a diff information if populating ExpOid and ExpDataType. Fields are exported for Marshaling purposes.
+type TestColumn struct {
+	Oid         uint32  `json:"oid"`
+	DataType    string  `json:"dataType"`
+	ExpOid      *uint32 `json:"expOid"`
+	ExpDataType *string `json:"expDataType"`
+}
+
+// testColumns maps column names with datatype description
+type testColumns map[string]*TestColumn
+
+// testTable have 2 use cases:
+// First: This is used to model pg_schema for postgres and cockroach db for comparing purposes by mapping tableNames
+// with columns.
+// Second: This is used to store and load expected diffs:
+// - Using it this way, a table name pointing to a zero length testColumns means that we expect this table to be missing
+//   in cockroach db
+// - If testColumns is not empty but columnName points to null, we expect that column to be missing in that table in
+//   cockroach db
+// - If column Name points to a not null TestColumn, the test column describes how we expect that data type to be
+//   different between cockroach db and postgres
+type testTable map[string]testColumns
+
+// addRow is used to load data from postgres or cockroach pg_catalog schema
+func (pgTables testTable) addRow(
+	tableName string, columnName string, dataType string, dataTypeOid uint32,
+) {
+	columns, ok := pgTables[tableName]
+
+	if !ok {
+		columns = make(testColumns)
+		pgTables[tableName] = columns
+	}
+
+	columns[columnName] = &TestColumn{dataTypeOid, dataType, nil, nil}
+}
+
+// addDiff is for the second use case for pgTables which objective is create a datatype diff
+func (pgTables testTable) addDiff(
```

Comment:
the parameter also should be renamed to `p` (and same for all the functions below)

Reply:
Done